File: app/src/main/java/com/example/demo/repository/BookRepository.java
package com.example.demo.repository;

import com.example.demo.model.Book;
import jakarta.persistence.LockModeType;

import java.util.List;
import java.util.Optional;

import jakarta.transaction.Transactional;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;

public interface BookRepository extends CrudRepository<Book, Long> {

    @Lock(LockModeType.OPTIMISTIC_FORCE_INCREMENT)
    Optional<Book> findWithLockingById(Long bookId);

    @Transactional
    @Query("SELECT b FROM Book b LEFT JOIN FETCH b.reviews WHERE b.title LIKE %:name%")
    List<Book> findBooksByTitleContaining(@Param("name") String name);
}

File: src/main/kotlin/com/example/demo/repository/BookRepository.kt
package com.example.demo.repository

import com.example.demo.model.Book
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.CrudRepository
import org.springframework.stereotype.Repository

@Repository
interface BookRepository : CrudRepository<Book, Long> {
//    abstract fun findByAuthorId(id: Long): List<Book>
    fun findByIdAndPageCount(id: Long, pageNumber: Int): Book?
    fun findAll(pageInfo: Pageable): MutableIterable<Book>
    abstract fun findByIdGreaterThan(id: Long, limit: Pageable): List<Book>
}